File: src/main/scala/io/github/fherbreteau/gatling/sftp/SftpDsl.scala
package io.github.fherbreteau.gatling.sftp

import io.gatling.core.config.GatlingConfiguration
import io.gatling.core.session.Expression
import io.github.fherbreteau.gatling.sftp.protocol.{SftpProtocol, SftpProtocolBuilder}

trait SftpDsl {

  def sftp(implicit configuration: GatlingConfiguration): SftpProtocolBuilder = SftpProtocolBuilder(configuration)

  def sftp(requestName: Expression[String]): Sftp = Sftp(requestName)

  implicit def sftpProtocolBuilder2SftpProtocol(builder: SftpProtocolBuilder): SftpProtocol = builder.build
}

File: src/main/java/io/github/fherbreteau/gatling/sftp/javaapi/SftpDsl.java
package io.github.fherbreteau.gatling.sftp.javaapi;

import io.gatling.javaapi.core.Session;
import io.gatling.javaapi.core.internal.Expressions;
import io.github.fherbreteau.gatling.sftp.javaapi.protocol.SftpProtocolBuilder;

import jakarta.annotation.Nonnull;
import java.util.function.Function;

/**
 * The entrypoint of the Gatling SFTP DSL
 */
public class SftpDsl {

    /**
     * Bootstrap a SFTP protocol configuration
     */
    public static final SftpProtocolBuilder sftp =
            new SftpProtocolBuilder(io.github.fherbreteau.gatling.sftp.protocol.SftpProtocolBuilder.apply(
                    io.gatling.core.Predef.configuration()));

    private SftpDsl() {
    }

    /**
     * Bootstrap a SFTP command configuration
     *
     * @param name the SFTP command name, expressed as a Gatling Expression Language String
     * @return the next DSL step
     */
    @Nonnull
    public static Sftp sftp(@Nonnull String name) {
        return new Sftp(Expressions.toStringExpression(name));
    }

    /**
     * Bootstrap a SFTP command configuration
     *
     * @param name the SFTP command name, expressed as a function
     * @return the next DSL step
     */
    @Nonnull
    public static Sftp sftp(@Nonnull Function<Session, String> name) {
        return new Sftp(Expressions.javaFunctionToExpression(name));
    }

}
